File: src/day16/Day16.kt
package day16

import java.io.File

const val FILENAME = "src/day16/Input.txt"

val input = File(FILENAME).readText().split("\n\n").toTypedArray()

val rules = input[0].split("\n").toTypedArray()

val list = input[2].split("\n")
val nearbyTickets = list.subList(1, list.size).toTypedArray()

val myTicket = input[1].split("\n")[1].split(",").map { it.toInt() }.toTypedArray()

fun part01(): Int {
    val listOfRanges = mutableListOf<(Int) -> Boolean>()

    for (rule in rules) {
        val reg = Regex(".*: ([0-9]*)-([0-9]*) or ([0-9]*)-([0-9]*)").find(rule)!!.groupValues
        val (r1, r2, r3, r4) = arrayOf(reg[1].toInt(), reg[2].toInt(), reg[3].toInt(), reg[4].toInt())
        listOfRanges.add(fun(num) = (r1 <= num && num <= r2) || (r3 <= num && num <= r4))
    }

    // Convert each string of comma separated values in nearbyTickets to an int[]
    // Then with errorRate initialized to 0
    // Take each array, select those numbers in the array that do not return true for even one of the ranges
    // Then with total initialized to errorRate, add each of them to total
    // Then errorRate <- total
    // Finally once each array in the list is exhausted, return errorRate
    return nearbyTickets.map { str -> str.split(",").map(Integer::parseInt).toTypedArray() }
        .fold(0, { errorRate, arr ->
            arr.filter { num -> !listOfRanges.any { fn -> fn(num) } }.fold(errorRate, { total, next -> total + next })
        })
}

fun part02(): Long {
    var departure = 1L

    val listOfRanges = mutableListOf<(Int) -> Boolean>()
    for (rule in rules) {
        val reg = Regex(".*: ([0-9]*)-([0-9]*) or ([0-9]*)-([0-9]*)").find(rule)!!.groupValues
        val (r1, r2, r3, r4) = arrayOf(reg[1].toInt(), reg[2].toInt(), reg[3].toInt(), reg[4].toInt())
        listOfRanges.add(fun(num) = (r1 <= num && num <= r2) || (r3 <= num && num <= r4))
    }
    val validTicketList = nearbyTickets.map {
        it.split(",").map(Integer::parseInt).toTypedArray()
    }.filter { arr -> arr.all { num -> listOfRanges.any { fn -> fn(num) } } }
    val validTickets2D = Array<IntArray>(validTicketList.size) { IntArray(validTicketList[0].size) }
    for (i in validTickets2D.indices) {
        for (j in validTickets2D[0].indices) {
            validTickets2D[i][j] = validTicketList[i][j]
        }
    }

    var ruleArrMap = mutableMapOf<String, BooleanArray>()
    for (rule in rules) {
        val reg = Regex(".*: ([0-9]*)-([0-9]*) or ([0-9]*)-([0-9]*)").find(rule)!!.groupValues
        val (r1, r2, r3, r4) = arrayOf(reg[1].toInt(), reg[2].toInt(), reg[3].toInt(), reg[4].toInt())

        val colArr = BooleanArray(validTickets2D[0].size)
        for (j in validTickets2D[0].indices) {
            var valid = true
            for (i in validTickets2D.indices) {
                if (!(validTickets2D[i][j] in r1..r2 || validTickets2D[i][j] in r3..r4)) {
                    valid = false
                    break
                }
            }
            if (valid) colArr[j] = true
        }
        ruleArrMap[reg[0]] = colArr
    }
    ruleArrMap = ruleArrMap.toSortedMap(
        compareBy { key -> ruleArrMap[key]!!.count { value -> value == true } }
    ).toMutableMap()

    val colIgnoreList = mutableListOf<Int>()
    for (key in ruleArrMap.keys) {
        val arr = ruleArrMap[key]!!
        for (col in colIgnoreList) arr[col] = false
        if (arr.count { it == true } == 1) {
            val index = arr.indexOf(true)
            if (key.startsWith("departure")) departure *= myTicket[index]
            colIgnoreList.add(index)
        }
    }

    return departure
}

fun main() {
    println("Advent of Code 2020, Day 16")
    println(part01())
    println(part02())
}

File: day16/day16.go
package day16

import (
	"advent-of-code-2020/utils"
	"fmt"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

const filename = "./day16/input.txt"

var input = strings.Split(utils.ReadString(filename), "\n\n")

var rules = strings.Split(input[0], "\n")

var list = strings.Split(input[2], "\n")
var nearbyTickets = list[1:]

var myTicket = utils.MapToInt(strings.Split(strings.Split(input[1], "\n")[1], ","))

func part01() int {
	listOfRanges := []func(int) bool{}

	for _, rule := range rules {
		reg := regexp.MustCompile(".*: ([0-9]*)-([0-9]*) or ([0-9]*)-([0-9]*)").FindStringSubmatch(rule)
		r1, _ := strconv.Atoi(reg[1])
		r2, _ := strconv.Atoi(reg[2])
		r3, _ := strconv.Atoi(reg[3])
		r4, _ := strconv.Atoi(reg[4])
		listOfRanges = append(listOfRanges, func(num int) bool {
			return ((num >= r1) && (num <= r2)) || ((num >= r3) && (num <= r4))
		})
	}

	errorRate := 0
	for _, nearbyTicket := range nearbyTickets {
		for _, n := range utils.MapToInt(strings.Split(nearbyTicket, ",")) {
			atLeastOneWithinRange := false
			for _, fn := range listOfRanges {
				if fn(n) {
					atLeastOneWithinRange = true
					break
				}
			}
			if !atLeastOneWithinRange {
				errorRate += n
			}
		}
	}
	return errorRate
}

func part02() int {
	departure := 1

	listOfRanges := []func(int) bool{}
	for _, rule := range rules {
		reg := regexp.MustCompile(".*: ([0-9]*)-([0-9]*) or ([0-9]*)-([0-9]*)").FindStringSubmatch(rule)
		r1, _ := strconv.Atoi(reg[1])
		r2, _ := strconv.Atoi(reg[2])
		r3, _ := strconv.Atoi(reg[3])
		r4, _ := strconv.Atoi(reg[4])
		listOfRanges = append(listOfRanges, func(num int) bool {
			return ((num >= r1) && (num <= r2)) || ((num >= r3) && (num <= r4))
		})
	}
	validTicketList := [][]int{}
	for _, nearbyTicket := range nearbyTickets {
		arr := utils.MapToInt(strings.Split(nearbyTicket, ","))
		allNumbersWithinSomeRange := true
		for _, n := range arr {
			atLeastOneWithinRange := false
			for _, fn := range listOfRanges {
				if fn(n) {
					atLeastOneWithinRange = true
					break
				}
			}
			if !atLeastOneWithinRange {
				allNumbersWithinSomeRange = false
				break
			}
		}
		if allNumbersWithinSomeRange {
			validTicketList = append(validTicketList, arr)
		}
	}
	validTickets2D := make([][]int, len(validTicketList))
	for i := range validTickets2D {
		validTickets2D[i] = make([]int, len(validTicketList[i]))
		for j := range validTickets2D[i] {
			validTickets2D[i][j] = validTicketList[i][j]
		}
	}

	ruleArrMap := make(map[string][]bool)
	for _, rule := range rules {
		reg := regexp.MustCompile(".*: ([0-9]*)-([0-9]*) or ([0-9]*)-([0-9]*)").FindStringSubmatch(rule)
		r1, _ := strconv.Atoi(reg[1])
		r2, _ := strconv.Atoi(reg[2])
		r3, _ := strconv.Atoi(reg[3])
		r4, _ := strconv.Atoi(reg[4])

		colArr := make([]bool, len(validTickets2D[0]))
		for j := range validTickets2D[0] {
			valid := true
			for i := range validTickets2D {
				if !(((validTickets2D[i][j] >= r1) && (validTickets2D[i][j] <= r2)) || ((validTickets2D[i][j] >= r3) && (validTickets2D[i][j] <= r4))) {
					valid = false
					break
				}
			}
			if valid {
				colArr[j] = true
			}
		}
		ruleArrMap[reg[0]] = colArr
	}
	ruleArrMapKeys := utils.MapKeys(ruleArrMap)
	slices.SortFunc(ruleArrMapKeys, func(a, b string) int {
		return utils.Count(ruleArrMap[a], true) - utils.Count(ruleArrMap[b], true)
	})

	colIgnoreList := []int{}
	for _, key := range ruleArrMapKeys {
		arr := ruleArrMap[key]
		for _, col := range colIgnoreList {
			arr[col] = false
		}
		if utils.Count(arr, true) == 1 {
			index := slices.Index(arr, true)
			if strings.HasPrefix(key, "departure") {
				departure *= myTicket[index]
			}
			colIgnoreList = append(colIgnoreList, index)
		}
	}

	return departure
}

func Main() {
	fmt.Println("Advent of Code 2020, Day 16")
	fmt.Println(part01())
	fmt.Println(part02())
}
